File: leetcode/ruby/search_in_rotated_sorted_list.rb
def search_in_rotated_sorted_list(nums, target)
  return -1 if nums.nil? || nums.empty?
  return 0 if nums.size == 1 && nums[0] == target
  return -1 if nums.size == 1 && nums[0] != target
  left = 0
  right = nums.size - 1
  while left <= right
    mid = (left + right) / 2
    if nums[mid] == target
      return mid
    elsif nums[mid] < target
      if nums[left] == target
        return left
      elsif nums[left] < target
        left = mid + 1
      else
        right = mid - 1
      end
    else
      if nums[right] == target
        return right
      elsif nums[right] > target
        right = mid - 1
      else
        left = mid + 1
      end
    end
  end
  return -1
end

// test
nums = [4, 5, 6, 7, 0, 1, 2]
target = 0
puts "nums: #{nums}, target: #{target}"
puts "result: #{search_in_rotated_sorted_list(nums, target)}"

File: leetcode/rust/search_in_rotated_sorted_list.rs
fn searchInRotatedSortedList(nums: Vec<i32>, target: i32) -> i32 {
    let mut l = 0;
    let mut r = nums.len() - 1;
    while l <= r {
        let m = (l + r) / 2;
        if nums[m] == target {
            return m as i32;
        }
        if nums[l] <= nums[m] {
            if nums[l] <= target && target < nums[m] {
                r = m - 1;
            } else {
                l = m + 1;
            }
        } else {
            if nums[m] < target && target <= nums[r] {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }
    }
    -1
}

fn main() {
    let nums = vec![4, 5, 6, 7, 0, 1, 2];
    let target = 0;
    let result = searchInRotatedSortedList(nums, target);
    println!("{}", result);
}